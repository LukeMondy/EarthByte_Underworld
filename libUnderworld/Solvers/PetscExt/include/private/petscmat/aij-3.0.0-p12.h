
#if !defined(__AIJ_H)
#define __AIJ_H

#include "private/matimpl.h"
/*  
    Struct header shared by SeqAIJ, SeqBAIJ and SeqSBAIJ matrix formats
*/
#define SEQAIJHEADER(datatype)	\
  PetscTruth        roworiented;      /* if true, row-oriented input, default */\
  PetscInt          nonew;            /* 1 don't add new nonzeros, -1 generate error on new */\
  PetscInt          nounused;         /* -1 generate error on unused space */\
  PetscTruth        singlemalloc;     /* if true a, i, and j have been obtained with one big malloc */\
  PetscInt          maxnz;            /* allocated nonzeros */\
  PetscInt          *imax;            /* maximum space allocated for each row */\
  PetscInt          *ilen;            /* actual length of each row */\
  PetscInt          reallocs;         /* number of mallocs done during MatSetValues() \
                                        as more values are set than were prealloced */\
  PetscInt          rmax;             /* max nonzeros in any row */\
  PetscTruth        keepzeroedrows;   /* keeps matrix structure same in calls to MatZeroRows()*/\
  PetscTruth        ignorezeroentries; \
  PetscInt          *xtoy,*xtoyB;     /* map nonzero pattern of X into Y's, used by MatAXPY() */\
  Mat               XtoY;             /* used by MatAXPY() */\
  PetscTruth        free_ij;          /* free the column indices j and row offsets i when the matrix is destroyed */ \
  PetscTruth        free_a;           /* free the numerical values when matrix is destroy */ \
  Mat_CompressedRow compressedrow;    /* use compressed row format */                      \
  PetscInt          nz;               /* nonzeros */                                       \
  PetscInt          *i;               /* pointer to beginning of each row */               \
  PetscInt          *j;               /* column values: j + i[k] - 1 is start of row k */  \
  PetscInt          *diag;            /* pointers to diagonal elements */                  \
  datatype          *a;               /* nonzero elements */                               \
  PetscScalar       *solve_work;      /* work space used in MatSolve */                    \
  IS                row, col, icol    /* index sets, used for reorderings */               

/*  
  MATSEQAIJ format - Compressed row storage (also called Yale sparse matrix
  format) or compressed sparse row (CSR).  The i[] and j[] arrays start at 0. For example,
  j[i[k]+p] is the pth column in row k.  Note that the diagonal
  matrix elements are stored with the rest of the nonzeros (not separately).
*/

/* Info about i-nodes (identical nodes) helper class for SeqAIJ */
typedef struct {
  MatScalar   *bdiag,*ibdiag;               /* diagonal blocks of matrix used for MatRelax_Inode() */
  PetscInt    bdiagsize;                       /* length of bdiag and ibdiag */
  PetscTruth  ibdiagvalid;                     /* do ibdiag[] and bdiag[] contain the most recent values */

  PetscTruth use;
  PetscInt   node_count;                    /* number of inodes */
  PetscInt   *size;                         /* size of each inode */
  PetscInt   limit;                         /* inode limit */
  PetscInt   max_limit;                     /* maximum supported inode limit */
  PetscTruth checked;                       /* if inodes have been checked for */
} Mat_Inode;

EXTERN PetscErrorCode MatView_Inode(Mat,PetscViewer);
EXTERN PetscErrorCode MatAssemblyEnd_Inode(Mat,MatAssemblyType);
EXTERN PetscErrorCode MatDestroy_Inode(Mat);
EXTERN PetscErrorCode MatCreate_Inode(Mat);
EXTERN PetscErrorCode MatSetOption_Inode(Mat,MatOption,PetscTruth);
EXTERN PetscErrorCode MatDuplicate_Inode(Mat,MatDuplicateOption,Mat*);
EXTERN PetscErrorCode MatILUDTFactor_Inode(Mat,IS,IS,const MatFactorInfo*,Mat*);
EXTERN PetscErrorCode MatLUFactorSymbolic_Inode(Mat,Mat,IS,IS,const MatFactorInfo*);
EXTERN PetscErrorCode MatILUFactorSymbolic_Inode(Mat,Mat,IS,IS,const MatFactorInfo*);


typedef struct {
  SEQAIJHEADER(MatScalar);
  Mat_Inode    inode;
  MatScalar    *saved_values;             /* location for stashing nonzero values of matrix */

  PetscScalar  *idiag,*mdiag,*ssor_work;  /* inverse of diagonal entries, diagonal values and workspace for Eisenstat trick */
  PetscTruth   idiagvalid;                     /* current idiag[] and mdiag[] are valid */
  PetscScalar  fshift,omega;                   /* last used omega and fshift */

  ISColoring   coloring;                  /* set with MatADSetColoring() used by MatADSetValues() */
} Mat_SeqAIJ;

/*
    Frees the a, i, and j arrays from the XAIJ (AIJ, BAIJ, and SBAIJ) matrix types
*/
#undef __FUNCT__  
#define __FUNCT__ "MatSeqXAIJFreeAIJ"
PETSC_STATIC_INLINE PetscErrorCode MatSeqXAIJFreeAIJ(Mat AA,MatScalar **a,PetscInt **j,PetscInt **i) 
{
                                     PetscErrorCode ierr;
                                     Mat_SeqAIJ     *A = (Mat_SeqAIJ*) AA->data;
                                     if (A->singlemalloc) {
                                       ierr = PetscFree3(*a,*j,*i);CHKERRQ(ierr);
                                     } else {
                                       if (A->free_a  && *a) {ierr = PetscFree(*a);CHKERRQ(ierr);}
                                       if (A->free_ij && *j) {ierr = PetscFree(*j);CHKERRQ(ierr);}
                                       if (A->free_ij && *i) {ierr = PetscFree(*i);CHKERRQ(ierr);}
                                     }
                                     *a = 0; *j = 0; *i = 0;
                                     return 0;
                                   }

/*
    Allocates larger a, i, and j arrays for the XAIJ (AIJ, BAIJ, and SBAIJ) matrix types
*/
#define MatSeqXAIJReallocateAIJ(Amat,AM,BS2,NROW,ROW,COL,RMAX,AA,AI,AJ,RP,AP,AIMAX,NONEW,datatype) \
  if (NROW >= RMAX) {\
	Mat_SeqAIJ *Ain = (Mat_SeqAIJ*)Amat->data;\
        /* there is no extra room in row, therefore enlarge */ \
        PetscInt   new_nz = AI[AM] + CHUNKSIZE,len,*new_i=0,*new_j=0; \
        datatype   *new_a; \
 \
        if (NONEW == -2) SETERRQ2(PETSC_ERR_ARG_OUTOFRANGE,"New nonzero at (%D,%D) caused a malloc",ROW,COL); \
        /* malloc new storage space */ \
        ierr = PetscMalloc3(BS2*new_nz,datatype,&new_a,new_nz,PetscInt,&new_j,AM+1,PetscInt,&new_i);CHKERRQ(ierr);\
 \
        /* copy over old data into new slots */ \
        for (ii=0; ii<ROW+1; ii++) {new_i[ii] = AI[ii];} \
        for (ii=ROW+1; ii<AM+1; ii++) {new_i[ii] = AI[ii]+CHUNKSIZE;} \
        ierr = PetscMemcpy(new_j,AJ,(AI[ROW]+NROW)*sizeof(PetscInt));CHKERRQ(ierr); \
        len = (new_nz - CHUNKSIZE - AI[ROW] - NROW); \
        ierr = PetscMemcpy(new_j+AI[ROW]+NROW+CHUNKSIZE,AJ+AI[ROW]+NROW,len*sizeof(PetscInt));CHKERRQ(ierr); \
        ierr = PetscMemcpy(new_a,AA,BS2*(AI[ROW]+NROW)*sizeof(datatype));CHKERRQ(ierr); \
        ierr = PetscMemzero(new_a+BS2*(AI[ROW]+NROW),BS2*CHUNKSIZE*sizeof(datatype));CHKERRQ(ierr);\
        ierr = PetscMemcpy(new_a+BS2*(AI[ROW]+NROW+CHUNKSIZE),AA+BS2*(AI[ROW]+NROW),BS2*len*sizeof(datatype));CHKERRQ(ierr);  \
        /* free up old matrix storage */ \
        ierr = MatSeqXAIJFreeAIJ(A,&Ain->a,&Ain->j,&Ain->i);CHKERRQ(ierr);\
        AA = new_a; \
        Ain->a = (MatScalar*) new_a;		   \
        AI = Ain->i = new_i; AJ = Ain->j = new_j;  \
        Ain->singlemalloc = PETSC_TRUE; \
 \
        RP          = AJ + AI[ROW]; AP = AA + BS2*AI[ROW]; \
        RMAX        = AIMAX[ROW] = AIMAX[ROW] + CHUNKSIZE; \
        Ain->maxnz += CHUNKSIZE; \
        Ain->reallocs++; \
      } \

EXTERN_C_BEGIN
EXTERN PetscErrorCode MatSeqAIJSetPreallocation_SeqAIJ(Mat,PetscInt,PetscInt*);
EXTERN_C_END
EXTERN PetscErrorCode MatILUFactorSymbolic_SeqAIJ(Mat,Mat,IS,IS,const MatFactorInfo*);
EXTERN PetscErrorCode MatICCFactorSymbolic_SeqAIJ(Mat,Mat,IS,const MatFactorInfo*);
EXTERN PetscErrorCode MatCholeskyFactorSymbolic_SeqAIJ(Mat,Mat,IS,const MatFactorInfo*);
EXTERN PetscErrorCode MatCholeskyFactorNumeric_SeqAIJ(Mat,Mat,const MatFactorInfo*);
EXTERN PetscErrorCode MatDuplicate_SeqAIJ(Mat,MatDuplicateOption,Mat*);
EXTERN PetscErrorCode MatMissingDiagonal_SeqAIJ(Mat,PetscTruth*,PetscInt*);
EXTERN PetscErrorCode MatMarkDiagonal_SeqAIJ(Mat);

EXTERN PetscErrorCode MatMult_SeqAIJ(Mat A,Vec,Vec);
EXTERN PetscErrorCode MatMultAdd_SeqAIJ(Mat A,Vec,Vec,Vec);
EXTERN PetscErrorCode MatMultTranspose_SeqAIJ(Mat A,Vec,Vec);
EXTERN PetscErrorCode MatMultTransposeAdd_SeqAIJ(Mat A,Vec,Vec,Vec);
EXTERN PetscErrorCode MatRelax_SeqAIJ(Mat,Vec,PetscReal,MatSORType,PetscReal,PetscInt,PetscInt,Vec);

EXTERN PetscErrorCode MatSetColoring_SeqAIJ(Mat,ISColoring);
EXTERN PetscErrorCode MatSetValuesAdic_SeqAIJ(Mat,void*);
EXTERN PetscErrorCode MatSetValuesAdifor_SeqAIJ(Mat,PetscInt,void*);

EXTERN PetscErrorCode MatGetSymbolicTranspose_SeqAIJ(Mat,PetscInt *[],PetscInt *[]);
EXTERN PetscErrorCode MatGetSymbolicTransposeReduced_SeqAIJ(Mat,PetscInt,PetscInt,PetscInt *[],PetscInt *[]);
EXTERN PetscErrorCode MatRestoreSymbolicTranspose_SeqAIJ(Mat,PetscInt *[],PetscInt *[]);
EXTERN PetscErrorCode MatToSymmetricIJ_SeqAIJ(PetscInt,PetscInt*,PetscInt*,PetscInt,PetscInt,PetscInt**,PetscInt**);
EXTERN PetscErrorCode MatLUFactorSymbolic_SeqAIJ(Mat,Mat,IS,IS,const MatFactorInfo*);
EXTERN PetscErrorCode MatLUFactorNumeric_SeqAIJ(Mat,Mat,const MatFactorInfo*);
EXTERN PetscErrorCode MatLUFactorNumeric_SeqAIJ_InplaceWithPerm(Mat,Mat,const MatFactorInfo*);
EXTERN PetscErrorCode MatLUFactor_SeqAIJ(Mat,IS,IS,const MatFactorInfo*);
EXTERN PetscErrorCode MatSolve_SeqAIJ(Mat,Vec,Vec);
EXTERN PetscErrorCode MatSolve_SeqAIJ_NaturalOrdering(Mat,Vec,Vec);
EXTERN PetscErrorCode MatSolve_SeqAIJ_InplaceWithPerm(Mat,Vec,Vec);
EXTERN PetscErrorCode MatSolveAdd_SeqAIJ(Mat,Vec,Vec,Vec);
EXTERN PetscErrorCode MatSolveTranspose_SeqAIJ(Mat,Vec,Vec);
EXTERN PetscErrorCode MatSolveTransposeAdd_SeqAIJ(Mat,Vec,Vec,Vec);
EXTERN PetscErrorCode MatMatSolve_SeqAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatEqual_SeqAIJ(Mat A,Mat B,PetscTruth* flg);
EXTERN PetscErrorCode MatFDColoringCreate_SeqAIJ(Mat,ISColoring,MatFDColoring);
EXTERN PetscErrorCode MatILUDTFactor_SeqAIJ(Mat,IS,IS,const MatFactorInfo*,Mat*);
EXTERN PetscErrorCode MatLoad_SeqAIJ(PetscViewer, const MatType,Mat*);
EXTERN PetscErrorCode RegisterApplyPtAPRoutines_Private(Mat);
EXTERN PetscErrorCode MatMatMult_SeqAIJ_SeqAIJ(Mat,Mat,MatReuse,PetscReal,Mat*);
EXTERN PetscErrorCode MatMatMultSymbolic_SeqAIJ_SeqAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatMatMultNumeric_SeqAIJ_SeqAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatPtAPSymbolic_SeqAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatPtAPNumeric_SeqAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatPtAPSymbolic_SeqAIJ_SeqAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatPtAPNumeric_SeqAIJ_SeqAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatMatMultTranspose_SeqAIJ_SeqAIJ(Mat,Mat,MatReuse,PetscReal,Mat*);
EXTERN PetscErrorCode MatMatMultTransposeSymbolic_SeqAIJ_SeqAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatMatMultTransposeNumeric_SeqAIJ_SeqAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatSetValues_SeqAIJ(Mat,PetscInt,const PetscInt[],PetscInt,const PetscInt[],const PetscScalar[],InsertMode);
EXTERN PetscErrorCode MatGetRow_SeqAIJ(Mat,PetscInt,PetscInt*,PetscInt**,PetscScalar**);
EXTERN PetscErrorCode MatRestoreRow_SeqAIJ(Mat,PetscInt,PetscInt*,PetscInt**,PetscScalar**);
EXTERN PetscErrorCode MatAXPY_SeqAIJ(Mat,PetscScalar,Mat,MatStructure);
EXTERN PetscErrorCode MatGetRowIJ_SeqAIJ(Mat,PetscInt,PetscTruth,PetscTruth,PetscInt*,PetscInt *[],PetscInt *[],PetscTruth *);
EXTERN PetscErrorCode MatRestoreRowIJ_SeqAIJ(Mat,PetscInt,PetscTruth,PetscTruth,PetscInt *,PetscInt *[],PetscInt *[],PetscTruth *);
EXTERN PetscErrorCode MatGetColumnIJ_SeqAIJ(Mat,PetscInt,PetscTruth,PetscTruth,PetscInt*,PetscInt *[],PetscInt *[],PetscTruth *);
EXTERN PetscErrorCode MatRestoreColumnIJ_SeqAIJ(Mat,PetscInt,PetscTruth,PetscTruth,PetscInt *,PetscInt *[],PetscInt *[],PetscTruth *);
EXTERN PetscErrorCode MatDestroy_SeqAIJ(Mat);
EXTERN PetscErrorCode MatView_SeqAIJ(Mat,PetscViewer);

EXTERN_C_BEGIN
EXTERN PetscErrorCode PETSCMAT_DLLEXPORT MatConvert_SeqAIJ_SeqSBAIJ(Mat,const MatType,MatReuse,Mat*);
EXTERN PetscErrorCode PETSCMAT_DLLEXPORT MatConvert_SeqAIJ_SeqBAIJ(Mat,const MatType,MatReuse,Mat*);
EXTERN PetscErrorCode PETSCMAT_DLLEXPORT MatConvert_SeqAIJ_SeqCSRPERM(Mat,const MatType,MatReuse,Mat*);
EXTERN PetscErrorCode PETSCMAT_DLLEXPORT MatReorderForNonzeroDiagonal_SeqAIJ(Mat,PetscReal,IS,IS);
EXTERN_C_END

#endif
