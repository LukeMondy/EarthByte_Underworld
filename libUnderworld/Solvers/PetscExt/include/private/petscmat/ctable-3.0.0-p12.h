/* Contributed by - Mark Adams */

#if !defined(__CTABLE300_H)
#define __CTABLE300_H

struct _n_PetscTable {
  PetscInt *keytable;
  PetscInt *table;
  PetscInt count;
  PetscInt tablesize;
  PetscInt head;
};

#include "petscctable.h"
#endif
