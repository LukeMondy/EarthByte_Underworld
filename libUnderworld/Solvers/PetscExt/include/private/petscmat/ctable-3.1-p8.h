/* Contributed by - Mark Adams */

#if !defined(__CTABLE31_H)
#define __CTABLE31_H

struct _n_PetscTable {
  PetscInt *keytable;
  PetscInt *table;
  PetscInt count;
  PetscInt tablesize;
  PetscInt head;
};

#include "petscctable.h"
#endif
