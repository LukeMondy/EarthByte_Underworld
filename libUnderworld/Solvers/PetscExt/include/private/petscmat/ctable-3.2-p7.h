/* Contributed by - Mark Adams */

#if !defined(__CTABLE_H)
#define __CTABLE_H

struct _n_PetscTable {
  PetscInt *keytable;
  PetscInt *table;
  PetscInt count;
  PetscInt tablesize;
  PetscInt head;
};

#include <petscctable.h>
#endif
