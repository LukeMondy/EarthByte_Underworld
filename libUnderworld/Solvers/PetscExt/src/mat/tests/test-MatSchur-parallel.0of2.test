EXEC 		= ./test-MatSchur-parallel.app
NPROCS 		= 2
TEST_TYPE       = tolerance_diff_pl,1.0e-10
