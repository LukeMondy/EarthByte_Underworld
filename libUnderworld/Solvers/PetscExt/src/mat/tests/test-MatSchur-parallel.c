/*@ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 **
 **    Copyright (c) 2007 - 2008, 
 **        Dave A. May [dave.mayhem23@gmail.com]
 **        School of Mathematical Sciences, 
 **        Monash University
 **        Victoria, 3800
 **        Australia
 **
 **    Copyright (c) 2009 - 2010, 
 **        Dave A. May [dave.mayhem23@gmail.com]
 **        Geophysical Fluid Dynamics, 
 **        Department of Earth Sciences,
 **        ETH ZÃ¼rich,
 **        Switzerland
 **
 **    Project:       PetscExt_v3.1
 **    Description:   A collection of extensions to the PETSc library.
 **
 **    All rights reserved.
 **    Redistribution and use in source and binary forms, with or without modification,
 **    are permitted provided that the following conditions are met:
 **
 **        * Redistributions of source code must retain the above copyright notice,
 **              this list of conditions and the following disclaimer.
 **        * Redistributions in binary form must reproduce the above copyright
 **              notice, this list of conditions and the following disclaimer in the
 **              documentation and/or other materials provided with the distribution.
 **        * Neither the name of the Monash University nor the names of its contributors
 **              may be used to endorse or promote products derived from this software
 **              without specific prior written permission.
 **
 **    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 **    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 **    THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 **    PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
 **    BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 **    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 **    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 **    HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 **    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT 
 **    OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 **
 **
 **    $Id$
 **
 ** ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~@*/

#include "stdio.h"
#include "stdlib.h"

#include <petsc.h>
#include <petscvec.h>
#include <petscmat.h>
#include <petscksp.h>
#include <petscpc.h>

#include "petscext_helpers.h"
#include "petscext_utils.h"
#include "petscext_vec.h"
#include "petscext_mat.h"


PetscErrorCode _MatView(Mat A,const char name[])
{
	PetscPrintf( PETSC_COMM_WORLD, "mat: %s\n", name );
	PetscViewerSetFormat( PETSC_VIEWER_STDOUT_WORLD, PETSC_VIEWER_ASCII_MATLAB_DENSE );
	MatView( A, PETSC_VIEWER_STDOUT_WORLD );
	PetscFunctionReturn(0);
}
PetscErrorCode _VecView(Vec x,const char name[])
{
	PetscPrintf( PETSC_COMM_WORLD, "vec: %s\n", name );
	VecView( x, PETSC_VIEWER_STDOUT_WORLD );
	PetscFunctionReturn(0);
}

/*
Build a non symmetric system with A22. Check MatMult, MatMultTranspose
*/
void test_schur_11_test2( void )
{
	Mat A_bk;
	Mat K,G,D,C;
	PetscInt M,N;
	Mat schur;
	Vec p, fhat;
	
	
	PetscPrintf( PETSC_COMM_WORLD, "Test: %s \n", __func__ );
	

	M = 7;
	N = 5;
	
	K = MATAIJIdentityNew( PETSC_COMM_WORLD, M );
	G = MATAIJNew( PETSC_COMM_WORLD, M,N, "G_" );
	D = MATAIJNew( PETSC_COMM_WORLD, N,M, "D_" );
	C = MATAIJNew( PETSC_COMM_WORLD, N,N, "C_" );
	
	MatFillStride( G, 2.0, 4.4 );
	MatFillStride( D, 5.0, 1.3 );
	MatFillStride( C, 1.0, 1.0 );
	
	
	MatCreate( PETSC_COMM_WORLD, &A_bk );
	MatSetSizes( A_bk, 2,2, 2,2 );
	MatSetType( A_bk, "block" );
	
	MatBlockSetValue( A_bk, 0,0, K,  DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	MatBlockSetValue( A_bk, 0,1, G,  DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	MatBlockSetValue( A_bk, 1,0, D,  DIFFERENT_NONZERO_PATTERN, INSERT_VALUES);
	MatBlockSetValue( A_bk, 1,1, C,  DIFFERENT_NONZERO_PATTERN, INSERT_VALUES);
	
	MatAssemblyBegin(A_bk,MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd(A_bk,MAT_FINAL_ASSEMBLY);
	Stg_MatDestroy( K );	Stg_MatDestroy( G );
	Stg_MatDestroy( D );	Stg_MatDestroy( C );
	
	
	MatCreateSchurFromBlock( A_bk, PETSC_NULL, "MatSchur_A11", &schur );
	MatGetVecs( schur, &p, &fhat );
	
	VecSet( p, 1.0 );
	
	MatAssemblyBegin( schur, MAT_FINAL_ASSEMBLY );
	MatAssemblyEnd( schur, MAT_FINAL_ASSEMBLY );
	
	MatMult( schur, p, fhat );
	_VecView(fhat,"S{1}");
	MatMultTranspose( schur, p, fhat );
	_VecView(fhat,"S^T{1}");

	
	
	Stg_MatDestroy( schur );
	
	Stg_VecDestroy( & p );
	Stg_VecDestroy( & fhat );
	Stg_MatDestroy( A_bk );
}

/*
Build a non symmetric system with A11 and A22. 
Set scale.
Set pc type lu
Check MatMult, MatMultTranspose
*/
void test_schur_22_test3( void )
{
	Mat A_bk;
	Mat K,G,D,C;
	PetscInt M,N;
	Mat schur;
	Vec u, f;
	KSP ksp;
	PC pc;
	
	PetscPrintf( PETSC_COMM_WORLD, "Test: %s \n", __func__ );

	M = 7;
	N = 5;
	
	K = MATAIJIdentityNew( PETSC_COMM_WORLD, M );
	G = MATAIJNew( PETSC_COMM_WORLD, M,N, "G_" );
	D = MATAIJNew( PETSC_COMM_WORLD, N,M, "D_" );
	C = MATAIJIdentityNew( PETSC_COMM_WORLD, N );
	
	MatFillStride( G, 2.0, 4.4 );
	MatFillStride( D, 5.0, 1.3 );
	MatScale( C, 2.0 );
	
	
	MatCreate( PETSC_COMM_WORLD, &A_bk );
	MatSetSizes( A_bk, 2,2, 2,2 );
	MatSetType( A_bk, "block" );
	
	MatBlockSetValue( A_bk, 0,0, K,  DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	MatBlockSetValue( A_bk, 0,1, G,  DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	MatBlockSetValue( A_bk, 1,0, D,  DIFFERENT_NONZERO_PATTERN, INSERT_VALUES);
	MatBlockSetValue( A_bk, 1,1, C,  DIFFERENT_NONZERO_PATTERN, INSERT_VALUES);
	
	MatAssemblyBegin(A_bk,MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd(A_bk,MAT_FINAL_ASSEMBLY);
	Stg_MatDestroy( K );	Stg_MatDestroy( G );
	Stg_MatDestroy( D );	Stg_MatDestroy( C );
	
	
	MatCreateSchurFromBlock( A_bk, 4.9, "MatSchur_A22", &schur );
	MatSchurGetKSP( schur, &ksp );
	
	MatAssemblyBegin( schur, MAT_FINAL_ASSEMBLY );
	MatAssemblyEnd( schur, MAT_FINAL_ASSEMBLY );
	
	MatGetVecs( schur, &u, &f );
	VecSet( u, 1.0 );
	
	MatMult( schur, u, f );
	_VecView(f,"S{1}");
	MatMultTranspose( schur, u, f );
	_VecView(f,"S^T{1}");
	

	{
		PetscInt M,N,m,n;
		PetscInt i,s,e;
		const PetscInt *r;
		PetscMPIInt np,rank;
		Mat S;

		S = schur;
	//	MatCreateNormal(K,&S);
	//	MatView( S, PETSC_VIEWER_STDOUT_WORLD );

		MPI_Comm_rank( PETSC_COMM_WORLD, &rank );
		MPI_Comm_size( PETSC_COMM_WORLD, &np );

		MatGetSize( S, &M,&N);
	//	printf("M,N = %d,%d \n", M,N);

		MatGetLocalSize( S, &m,&n );
	//	printf("m,n = %d,%d \n", m,n );

		MatGetOwnershipRange( S, &s,&e);
	//	printf("s,e = %d,%d \n", s,e );

	//	MatGetOwnershipRanges( S, &r );
	//	for( i=0; i<np+1; i++ ) {
	//		printf("r[%d]=%d \n", i, r[i] );
	//	}
	}
	
	
	Stg_MatDestroy( schur );	
	Stg_VecDestroy( & u );
	Stg_VecDestroy( & f );
	Stg_MatDestroy( A_bk );
}



int main(int argc,char **argv)
{
	int i;
	Vec x;
	
	PetscInitialize(&argc,&argv,(char *)0,PETSC_NULL);
	PetscExtVecRegisterAll();
	PetscExtMatRegisterAll();

	
	for( i=0; i<1; i++ ) {
		test_schur_11_test2();PetscPrintf(PETSC_COMM_WORLD,"\n");
		test_schur_22_test3();PetscPrintf(PETSC_COMM_WORLD,"\n");
	}	

        
	PetscExtVecRegisterDestroyAll();
	PetscExtMatRegisterDestroyAll();
	PetscFinalize();
	return 0;
}
