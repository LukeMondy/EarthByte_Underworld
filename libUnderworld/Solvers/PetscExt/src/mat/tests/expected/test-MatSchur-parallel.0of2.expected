Test: test_schur_11_test2 
vec: S{1}
Process [0]
27912.2
52348
76783.8
Process [1]
101220
125655
vec: S^T{1}
Process [0]
68447
72615.4
76783.8
Process [1]
80952.2
85120.6

Test: test_schur_22_test3 
vec: S{1}
Process [0]
31959.3
83083.4
134208
185332
Process [1]
236456
287580
338704
vec: S^T{1}
Process [0]
159648
168209
176770
185332
Process [1]
193893
202454
211016

